File: AspNet472CSharpMVC5/Controllers/HomeController.cs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.

using System.Web;
using System.Web.Mvc;
using System.Web.SessionState;

using Microsoft.Owin.Security;
using Microsoft.Owin.Security.Cookies;
using System.Security.Claims;

namespace AspNet472CSharpMVC5.Controllers
{
    [SessionState(SessionStateBehavior.Required)]
    public class HomeController : Controller
    {
        public ActionResult Index()
        {
            return View();
        }

        [HttpPost]
        public ActionResult CreateCookies()
        {
            const string CookieName = "sampleCookie";
            if (Request.Cookies[CookieName] == null)
            {
                // Create the cookie
                HttpCookie sameSiteCookie = new HttpCookie(CookieName);

                // Set a value for the cookie
                sameSiteCookie.Value = "sample";

                // Set the secure flag, which Chrome's changes will require for SameSite none.
                // Note this will also require you to be running on HTTPS
                sameSiteCookie.Secure = true;

                // Set the cookie to HTTP only which is good practice unless you really do need
                // to access it client side in scripts.
                sameSiteCookie.HttpOnly = true;

                // Add the SameSite attribute
                sameSiteCookie.SameSite = SameSiteMode.Lax;

                // Add the cookie to the response cookie collection
                Response.Cookies.Add(sameSiteCookie);
            }

            // Create a session variable to get a session cookie created.
            Session["sample"] = "sample";

            // And fake a login to create a membership cookie with OWIN cookie authentication.
            // Note we're not setting any sameSite attributes here.
            var claims = new[] { new Claim(ClaimTypes.NameIdentifier, "username")};
            var identity = new ClaimsIdentity(claims, CookieAuthenticationDefaults.AuthenticationType);
            var ctx = HttpContext.GetOwinContext();
            var authManager = ctx.Authentication;
            authManager.SignIn(new AuthenticationProperties { IsPersistent = false}, identity);

            return RedirectToAction("Index");
        }
    }
}

File: AspNet472VisualBasicMVC5/Controllers/HomeController.vb
' Copyright (c) .NET Foundation. All rights reserved.
' Licensed under the Apache License, Version 2.0. See License.txt In the project root for license information.

Imports System.Security.Claims
Imports Microsoft.Owin.Security
Imports Microsoft.Owin.Security.Cookies

Public Class HomeController
    Inherits System.Web.Mvc.Controller

    Function Index() As ActionResult
        Return View()
    End Function

    <HttpPost()>
    Function CreateCookies() As ActionResult
        Const CookieName = "sampleCookie"

        If Request.Cookies(CookieName) Is Nothing Then

            ' Create the cookie
            Dim sameSiteCookie As New HttpCookie(CookieName)

            ' Set a value for the cookie
            sameSiteCookie.Value = "sample"

            ' Set the secure flag, which Chrome's changes will require for SameSite none.
            ' Note this will also require you to be running on HTTPS
            sameSiteCookie.Secure = True

            ' Set the cookie to HTTP only which is good practice unless you really do need
            ' to access it client side in scripts.
            sameSiteCookie.HttpOnly = True

            ' Expire the cookie in 1 minute
            sameSiteCookie.Expires = Date.Now.AddMinutes(1)

            ' Add the SameSite attribute, this will emit the attribute with a value of none.
            ' To Not emit the attribute at all set the SameSite property to -1.
            sameSiteCookie.SameSite = SameSiteMode.None

            ' Add the cookie to the response cookie collection
            Response.Cookies.Add(sameSiteCookie)
        End If

        ' Create a session variable to get a session cookie created.
        Session("sample") = "sample"

        ' And fake a login to create a membership cookie with OWIN cookie authentication.
        ' Note we're not setting any sameSite attributes here.
        Dim claims = {New Claim(ClaimTypes.NameIdentifier, "username")}
        Dim identity = New ClaimsIdentity(claims, CookieAuthenticationDefaults.AuthenticationType)
        Dim ctx = HttpContext.GetOwinContext()
        Dim authManager = ctx.Authentication
        authManager.SignIn(New AuthenticationProperties With {.IsPersistent = False}, identity)

        Return RedirectToAction("Index")
    End Function
End Class
